eos-updater/apply.c

```diff
@@ -96,6 +96,85 @@ get_test_osname (void)
   return g_getenv ("EOS_UPDATER_TEST_UPDATER_OSTREE_OSNAME");
 }
 
+static gboolean
+update_remote_collection_id (OstreeRepo   *repo,
+                             const gchar  *refspec,
+                             GError      **error)
+{
+  g_autofree gchar *remote = NULL;
+  g_autofree gchar *ref = NULL;
+  g_autofree gchar *cur_collection_id = NULL;
+  g_autofree gchar *checksum = NULL;
+  g_autofree gchar *remote_group = NULL;
+  g_autoptr(GVariant) commit = NULL;
+  g_autoptr(GVariant) metadata = NULL;
+  g_autoptr(GVariant) collection_id_v = NULL;
+  g_autoptr(GKeyFile) config = NULL;
+  const char *collection_id;
+
+  if (!ostree_parse_refspec (refspec, &remote, &ref, error))
+    return FALSE;
+  if (remote == NULL)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                   "Invalid refspec ‘%s’ in origin: did not contain a remote name",
+                   refspec);
+      return FALSE;
+    }
+
+  /* If the remote already has a collection ID set, there's nothing to do */
+  if (!ostree_repo_get_remote_option (repo, remote, "collection-id", NULL,
+                                      &cur_collection_id, error))
+    return FALSE;
+  if (cur_collection_id != NULL && *cur_collection_id != '\0')
+    return TRUE;
+
+  if (!ostree_repo_resolve_rev (repo, refspec, FALSE, &checksum, error))
+    return FALSE;
+  if (!ostree_repo_load_variant (repo,
+                                 OSTREE_OBJECT_TYPE_COMMIT,
+                                 checksum,
+                                 &commit,
+                                 error))
+    return FALSE;
+
+  metadata = g_variant_get_child_value (commit, 0);
+  collection_id_v = g_variant_lookup_value (metadata, "ostree.collection-binding", G_VARIANT_TYPE_STRING);
+  if (collection_id_v == NULL ||
+      (collection_id = g_variant_get_string (collection_id_v, NULL)) == NULL)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to read ostree.collection-binding metadata key");
+      return FALSE;
+    }
```

Comment:
Would it be better to look at the collection ID in the summary file (since that will definitely match what the remote specifies) rather than the binding in the commit?

Reply:
How could the one in the commit not be the correct one? I don't think we can fetch the summary here in the apply stage, which I think is supposed to work offline